internal/repo/pg_repo.go

```diff
@@ -0,0 +1,173 @@
+package repo
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/google/uuid"
+	"github.com/ozonva/ova-service-api/internal/models"
+	"github.com/rs/zerolog/log"
+
+	_ "github.com/jackc/pgx/v4"
+	_ "github.com/jackc/pgx/v4/stdlib"
+)
+
+type PostgresServiceRepo struct {
+	ctx context.Context
+	db  *sql.DB
+}
+
+func NewPostgresServiceRepo(ctx context.Context, dsn string) (*PostgresServiceRepo, error) {
+	db, err := sql.Open("pgx", dsn)
+
+	if err != nil {
+		log.Err(err).Msg("Can't load pgx driver")
+		return nil, err
+	}
+
+	if connErr := db.PingContext(ctx); connErr != nil {
+		log.Err(err).Msg("Failed to connect to database")
+		return nil, err
+	}
+
+	return &PostgresServiceRepo{
+		ctx: ctx,
+		db:  db,
+	}, nil
+}
+
+func (repo *PostgresServiceRepo) AddServices(services []models.Service) error {
+	log.Debug().Msg("PostgresServiceRepo.AddServices call")
+
+	if len(services) == 0 {
+		return nil
+	}
+
+	tx, err := repo.db.BeginTx(repo.ctx, nil)
+	defer func(tx *sql.Tx) {
+		txErr := tx.Rollback()
+		if txErr != sql.ErrTxDone {
+			log.Err(txErr).Msg("Can' rollback the transaction")
+		}
+	}(tx)
+
+	if err != nil {
+		log.Err(err).Msg("Failed to begin transaction")
+		return err
+	}
+
+	query := `INSERT INTO services (id, user_id, description, service_name, service_address, when_local, when_utc)
+			VALUES ($1, $2, $3, $4, $5, $6, $7)`
+
+	for _, service := range services {
+		if _, err = tx.ExecContext(repo.ctx, query, service.ID, service.UserID, service.Description, service.ServiceName, service.ServiceAddress, service.WhenLocal, service.WhenUTC); err != nil {
+			log.Err(err).Msg("Failed to begin transaction")
+			return err
+		}
+	}
+
+	if err = tx.Commit(); err != nil {
+		log.Err(err).Msg("Error occurred during commit transaction")
+		return err
+	}
```

Comment:
Лучше одним INSERT'ом

Reply:
Исправил, но получилось страшновато:

```go
	for i, service := range services {
		// This looks crazy, but this is because of placeholders in pgx: $1, $2, $3, etc. I can't find a way how to do it easier using database/sql.
		queryParts[i+1] = fmt.Sprintf("($%d, $%d, $%d, $%d, $%d, $%d, $%d),", 7*i+1, 7*i+2, 7*i+3, 7*i+4, 7*i+5, 7*i+6, 7*i+7)
		queryValues = append(queryValues, service.ID, service.UserID, service.Description, service.ServiceName, service.ServiceAddress, service.WhenLocal, service.WhenUTC)
	}
```

Так, потому что нельзя использовать `?` в качестве placeholder. Обойти можно через [pgx.CopyFrom](https://pkg.go.dev/github.com/jackc/pgx#hdr-Copy_Protocol), но мне очень хотелось оставать в поле возможностей `database/sql`, и более простого решения я не нашел.

Reply:
Есть вот такой вот пакет https://github.com/huandu/go-sqlbuilder , чтобы удобно писать запросы
После составления запроса можно заменить ? на нумерованные placeholderы при помощи sqlx.Rebind

Reply:
Поправил, получилось намного аккуратнее, правда. В целом, наверное, можно было бы обойтись только sqlx.Rebind, но с точки зрения читаемости с билдером совсем хорошо получается.